File: QA-Assignment/BusinessWorkflowAutomation/pages/knowledge_base_page.py
import re
from playwright.sync_api import Page, expect
from pages.base_page import BasePage


class KnowledgeBasePage(BasePage):
    """Page Object Model for Knowledge Base Management (/workspace/index)."""

    def __init__(self, page: Page):
        super().__init__(page)
        self.page_heading = page.get_by_role("heading", name=re.compile(r"Add a website", re.IGNORECASE))
        self.url_input = page.get_by_placeholder("https://example.com")
        self.initial_question_input = page.get_by_placeholder("What is this website about?")
        self.max_pages_select = page.locator("select")
        self.submit_button = page.get_by_role("button", name=re.compile(r"Crawl and index website", re.IGNORECASE))

    def _get_delete_modal(self):
        """Get the scoped delete confirmation modal to avoid strict mode violation."""
        return self.page.locator("div").filter(
            has=self.page.get_by_role("heading", name=re.compile(r"Delete Knowledge Base", re.IGNORECASE))
        ).first

    def goto(self):
        """Navigate to Add Website / Knowledge Base indexing page."""
        self.navigate_to("/workspace/index")
        expect(self.page_heading).to_be_visible(timeout=15000)

    def add_website(self, url: str, initial_question: str = "", max_pages: int = 1):
        """Fill and submit website indexing form."""
        expect(self.url_input).to_be_visible(timeout=15000)
        self.url_input.fill(url)
        if initial_question:
            self.initial_question_input.fill(initial_question)
        if max_pages:
            self.max_pages_select.select_option(str(max_pages))
        self.submit_button.click()

    def assert_website_indexed(self, domain_or_url: str):
        """Assert website is indexed and listed in the Knowledge Base list.
        Waits up to 60s for crawling/indexing to complete on deployed infra.
        """
        site_item = self.page.get_by_text(domain_or_url, exact=False)
        expect(site_item.first).to_be_visible(timeout=60000)

    def delete_website(self, domain_or_url: str):
        """Delete an indexed website from Knowledge Base list."""
        # Navigate to /workspace/index if not already there
        current_url = self.page.url  # property, no parentheses
        if "/workspace/index" not in current_url:
            self.navigate_to("/workspace/index")
            expect(self.page_heading).to_be_visible(timeout=15000)

        site_container = self.page.locator("div", has_text=domain_or_url).filter(
            has=self.page.locator('button[title="Delete website"]')
        ).first
        site_container.hover()

        delete_btn = site_container.locator('button[title="Delete website"]').first
        delete_btn.click()

        # Wait for modal and click the scoped "Delete" button inside the modal
        modal = self._get_delete_modal()
        expect(modal).to_be_visible(timeout=8000)
        confirm_btn = modal.get_by_role("button", name="Delete", exact=True)
        expect(confirm_btn).to_be_visible(timeout=5000)
        confirm_btn.click()

    def assert_website_deleted(self, domain_or_url: str):
        """Assert website is removed from Knowledge Base list."""
        site_item = self.page.get_by_text(domain_or_url, exact=True)
        expect(site_item).to_have_count(0, timeout=15000)

File: QA-Assignment/Automation/pages/KnowledgeBasePage.js
import { expect } from '@playwright/test';

/**
 * Page Object Model for Knowledge Base / Website Management (/workspace/index)
 */
export class KnowledgeBasePage {
  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;

    // Locators
    this.pageHeading = page.getByRole('heading', { name: /Add a website/i });
    this.urlInput = page.getByPlaceholder('https://example.com');
    this.initialQuestionInput = page.getByPlaceholder('What is this website about?');
    this.maxPagesSelect = page.locator('select');
    this.submitButton = page.getByRole('button', { name: /Crawl and index website/i });
    // Scope modal delete button to the modal container to avoid strict mode violation
    // The modal contains heading "Delete Knowledge Base"
    this.deleteModal = page.locator('div').filter({ has: page.getByRole('heading', { name: /Delete Knowledge Base/i }) }).first();
    this.modalDeleteButton = this.deleteModal.getByRole('button', { name: 'Delete', exact: true });
    this.modalCancelButton = this.deleteModal.getByRole('button', { name: 'Cancel' });
  }

  /**
   * Navigate to Add Website page
   */
  async goto() {
    await this.page.goto('/workspace/index');
    await expect(this.pageHeading).toBeVisible({ timeout: 15000 });
  }

  /**
   * Submit website indexing form
   * @param {string} url
   * @param {string} [initialQuestion='']
   * @param {number|string} [maxPages=1]
   */
  async addWebsite(url, initialQuestion = '', maxPages = 1) {
    await expect(this.urlInput).toBeVisible({ timeout: 15000 });
    await this.urlInput.fill(url);
    if (initialQuestion) {
      await this.initialQuestionInput.fill(initialQuestion);
    }
    if (maxPages) {
      await this.maxPagesSelect.selectOption(String(maxPages));
    }
    await this.submitButton.click();
  }

  /**
   * Assert website is indexed and listed in the Knowledge Base list.
   * Allows up to 60s for the deployed crawler to complete indexing.
   * @param {string} domainOrUrl
   */
  async assertWebsiteIndexed(domainOrUrl) {
    const siteItem = this.page.getByText(domainOrUrl, { exact: false });
    await expect(siteItem.first()).toBeVisible({ timeout: 60000 });
  }

  /**
   * Delete website from Knowledge Base
   * @param {string} domainOrUrl
   */
  async deleteWebsite(domainOrUrl) {
    // Navigate to /workspace/index if not already there
    const currentUrl = this.page.url();
    if (!currentUrl.includes('/workspace/index')) {
      await this.page.goto('/workspace/index');
      await expect(this.pageHeading).toBeVisible({ timeout: 15000 });
    }

    const siteContainer = this.page.locator('div', { hasText: domainOrUrl }).filter({
      has: this.page.locator('button[title="Delete website"]'),
    }).first();
    await siteContainer.hover();

    const deleteBtn = siteContainer.locator('button[title="Delete website"]').first();
    await deleteBtn.click();

    // Wait for the modal to appear, then click the scoped Delete button
    const deleteModal = this.page.locator('div').filter({
      has: this.page.getByRole('heading', { name: /Delete Knowledge Base/i }),
    }).first();
    await expect(deleteModal).toBeVisible({ timeout: 8000 });

    const confirmDeleteBtn = deleteModal.getByRole('button', { name: 'Delete', exact: true });
    await expect(confirmDeleteBtn).toBeVisible({ timeout: 5000 });
    await confirmDeleteBtn.click();
  }

  /**
   * Assert website is removed from Knowledge Base list
   * @param {string} domainOrUrl
   */
  async assertWebsiteDeleted(domainOrUrl) {
    const siteItem = this.page.getByText(domainOrUrl, { exact: true });
    await expect(siteItem).toHaveCount(0, { timeout: 15000 });
  }
}
